### ISSUE TITLE
Recommend async alternatives to blocking IO calls

### ISSUE THREAD

**Issue by Zac-HD** [Zac-HD/flake8-trio#58]
### Background

This is going to have a fair bit of overlap with https://github.com/cooperlees/flake8-async/pull/6 and https://github.com/cooperlees/flake8-async/pull/9, precisely because I hoped for a framework-agnostic approach.  However, I've ultimately changed my mind because Trio-specific checks can be more specific, and apply to cases where there are no generic warnings or recommendations.  Using `trio.wrap_file(os.fdopen(..))` is an example of both!

### Specific recommendations

I think each cluster of these should probably use a separate `TRIO2xx` code, so they can be individually suppressed - it's useful for incremental adoption at least.  Let's reserve `TRIO200` for user-supplied recommendations (see below).

* [x] `HTTP` requests: #94 
  - `requests` and sync-`httpx` - as for `flake8-async`
  - `urllib3.request()`, but not bare `request()` (too generic) - use equivalent method with an `httpx.AsyncClient`
  - `urllib.request.urlopen()` (stdlib) - likewise, including `request.urlopen()` and `urlopen()`
  - [x] httpx.Client() #118

* [x] Process invocations: #94
  - `subprocess.Popen` and `os.popen` - `await nursery.start(trio.run_process, ...)`
  - other `subprocess` functions, `os.system` - `await trio.run_process(...)`
  - `os.spawn[vl]p?e?` family - usually as for system; if `mode=` is given and not `[os.]P_WAIT` then as for popen.

File IO:
- [x] Use `trio.open_file()` instead of `open()`, `io.open`, `io.open_code`, or `os.fdopen` in async functions #101
- [x] Use of blocking methods on file objects - wrap in `trio.wrap_file()` to get an *async* file object (see below)
- [x] Use of blocking methods on `pathlib.Path` objects - wrap in `trio.Path()` to get an async path object (again, see below) #104 #111

Path manipulation:
- [x] Any use of `os.path` module - prefer using `trio.Path` objects (it's a copout but good advice IMO) #100
- [ ] TBD: `os.walk`, `os.pidfd_open` - don't have exact equivalents on `trio.Path`? 🤔 

OS functionality:
- `os.wait.*()` (except `waitstatus_to_exitcode`) - must be wrapped in `trio.to_thread.run_sync()`
- `os.pidfd_open` - must be wrapped in `trio.wrap_file()` immediately
- `os.remove`, `os.unlink` - prefer `trio.Path` methods or wrap in `trio.to_thread.run_sync()`
- [ ] ...you know what there's too much to list in the `os` module and it's rare in async code, let's revisit it later.

I haven't observed antipatterns for use of the `socket` library in sync code and don't want to speculate, but that's one of several potential future categories here, along with various websocket libraries.  Note also that this is large enough that it should probably be split into multiple PRs, and leave anything nontrivial for after we get the basics in.

### The problem with Path (and file) objects
* [x] ...

is that they're often passed in as arguments to a function, so all the linter can see is `varname.read()` - and we can't warn on that without an unacceptable level of false alarms.  Of course warning on `Path("log.txt").write_text(data)` is reasonable, but I think we can do better by checking for arguments which are explicitly annotated as (*non-Trio!*) `Path`s and warning on those.  This should be left to a future PR, I'm just writing it up here so I can see all the design considerations in one place.

### Config option to add simple recommendations 
* [x] #80
It'd be nice if end users could check for blocking uses of nonpublic code, and also nice to try out new warnings without needing to ship an updated version.  To that end, let's make each collection of things-to-check extensible where that's easy-ish to do; but to ease maintainence we'll only support configuring the simple patterns for now:

```ini
# Warn if we see the thing before ` -> `, and recommend the thing after as a literal string.
# Error out on any entry which doesn't have that delimiter.

trio200-blocking-calls =
    # If you want to support with-or-without library name, list each case.
    # It's verbose but a simple implementation is worth a lot!
    bar -> qux.bar,
    foo.bar -> qux.bar,
    
	# Later: syntax for 'method call' warnings.   (literal `(...)` - empty parens will be an error)
    CustomPath(...).read_text -> AsyncCustomPath(...).read_text
```

* [ ] The `(...)` method-call syntax should also be used from type annotations once we get that working.  Complicated logic like "`trio.wrap_file(os.fdopen(...))` is OK" just isn't user-configurable.

**Comment by jakkdl** [comment]
> ```
> # If you want to support with-or-without library name, list each case.
> # It's verbose but a simple implementation is worth a lot!```

It's actually supporting recursive dotting that's the most complicated part, since that requires recursively stepping through the attribute structure and building up the name, so supporting wildcard isn't gonna complicate it more at all. If anything explicitly not supporting it here means I add a parameter to `fnmatch_qualified_name` and making that function more complicated :P
But at least until Path/file objects are implemented it feels like a user should be able to configure `*.read` if they really want to. If you don't want to use wildcards you can just refrain from configuring that, or I suppose one could have a flag to enforce never using `*`.

**Comment by jakkdl** [comment]
Possible future addition to TRIO211 - track the variable names of `httpx.Client()` and urllib3 pools so the following code can be warned on w/o excessive false alarms:
```python
client = httpx.Client()
client.get('https://example.org')
```
(Adding it here so it's not forgotten, from discussion in #94)

**Comment by jakkdl** [comment]
Should `os.fdopen(...)` be replaced with `trio.open_file(...)` or `trio.wrap_file(os.fdopen(...))`? In the bullet list you say the former, but in the intro you say the latter.

**Comment by jakkdl** [comment]
Is there any way to do
> Use of blocking methods on file objects - wrap in trio.wrap_file() to get an async file object (see below)
Use of blocking methods on pathlib.Path objects - wrap in trio.Path() to get an async path object (again, see below)

without tracking variable names? If I'm reading the documentation correctly async file objects have the same interface as sync file objects - so can't do a blacklist of of method names.

**Comment by jakkdl** [comment]
I think I have open PR's for all the simple warnings now, I reformatted the initial post so it was easier to see what is done and what's left.

**Comment by jakkdl** [comment]
Prompted by https://github.com/Zac-HD/flake8-trio/pull/103#discussion_r1088440075

The following os functions are checked by flake8-async, and not yet by us:
```python
"posix_spawn",
"posix_spawnp",
"wait",
"wait3",
"wait4",
"waitid",
"waitpid",
```

So might be time to revisit the os module, or can just implement these ones specifically so that flake8-asyncio is fully redundant. (and once that's done, update `__init__.py` to remove the reference)

**Comment by Zac-HD** [comment]
Let's just do those ones to start with, and revisit the rest later 👍

**Comment by jakkdl** [comment]
> Let's just do those ones to start with, and revisit the rest later +1

Should [`os.posix_spawn[p]`](https://docs.python.org/3/library/os.html#os.posix_spawn) be replaced by `await nursery.start(trio.run_proces, ...)` or `await trio.run_process(...)` ?

And what's the replacement for the [`wait`](https://docs.python.org/3/library/os.html#os.wait) family?

**Comment by jakkdl** [comment]
Updated the first post.
User-configured method-call warnings should be relatively straightforward after #118, otherwise I'm starting to be done with everything that's been specified.

**Comment by Zac-HD** [comment]
I don't actually have a use for user-configured method call warnings at the moment and that sounds pretty complicated, so let's call this one complete as it stands - the remaining `os.___` stuff is also rare, and I'd rather get #61 for the benefit of other OSS packages 🙂

**Comment by jakkdl** [comment]
With `TypeTrackerVisitor200` implementing user-configured method call warnings isn't going to be hard, it's mostly just about parsing the user-specified data and to what complexity. But that's probably better done once there's a need for it.
I'll make a final PR with the `os` functions warned about in flake8-async so it's fully redundant, but otherwise :partying_face:
